src/util/fields.py

```diff
@@ -510,49 +510,60 @@ def rename_field(cr, model, old, new, update_references=True, domain_adapter=Non
         # skip all inherit, they will be handled by the recursive call
         update_field_usage(cr, model, old, new, domain_adapter=domain_adapter, skip_inherit="*")
 
-    try:
-        with savepoint(cr):
-            cr.execute("UPDATE ir_model_fields SET name=%s WHERE model=%s AND name=%s RETURNING id", (new, model, old))
-            [fid] = cr.fetchone() or [None]
-    except psycopg2.IntegrityError:
+    # search for an existing custom field.
+    cr.execute("SELECT 1 FROM ir_model_fields WHERE model = %s AND name = %s", [model, new])
+    if cr.rowcount:
         # If a field with the same name already exists for this model (e.g. due to a custom module),
         # rename it to avoid clashing and warn the customer
         custom_name = new + "_custom"
         rename_field(cr, model, new, custom_name, update_references, domain_adapter=None, skip_inherit=skip_inherit)
-        cr.execute("UPDATE ir_model_fields SET name=%s WHERE model=%s AND name=%s RETURNING id", (new, model, old))
-        [fid] = cr.fetchone() or [None]
         msg = (
             "The field %r from the model %r is now a standard Odoo field, but it already existed in the database "
             "(coming from a non-standard module) and thus has been renamed to %r" % (new, model, custom_name)
         )
         add_to_migration_reports(msg, "Non-standard fields")
 
+    cr.execute("UPDATE ir_model_fields SET name=%s WHERE model=%s AND name=%s RETURNING id", (new, model, old))
+    [fid] = cr.fetchone() or [None]
+
     if fid:
+        # recreate the xmlids
         name = IMD_FIELD_PATTERN % (model.replace(".", "_"), new)
-        # In some cases the same field may be present on ir_model_data with both the double __ and single _ name
-        # version. To avoid conflicts (module, name) on the UPDATE below we keep only the double __ version
+        # duplicate key. May happen du to conflict between
+        # some_model.sub_id and some_model_sub.id
+        # (before saas~11.2, where pattern changed)
```

Comment:
Some typos. I'd make this clear enough to avoid people figuring out what pattern we are talking about.
```suggestion
        # Duplicated key. May happen due to some_model.sub_id and some_model_sub.id
        # both leading to some_model_sub_id xmlid before saas~11.2 where the pattern
        # was fixed to be some_module__sub_id and some_model_sub__id respectively.
```